pkg/registry/flowcontrol/rest/storage_flowcontrol.go

```diff
@@ -0,0 +1,147 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/registry/generic"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	serverstorage "k8s.io/apiserver/pkg/server/storage"
+	flowcontrolbootstrap "k8s.io/apiserver/pkg/util/flowcontrol/bootstrap"
+	flowcontrolclient "k8s.io/client-go/kubernetes/typed/flowcontrol/v1alpha1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+	flowschemastore "k8s.io/kubernetes/pkg/registry/flowcontrol/flowschema/storage"
+	prioritylevelconfigurationstore "k8s.io/kubernetes/pkg/registry/flowcontrol/prioritylevelconfiguration/storage"
+)
+
+const PostStartHookName = "apiserver/bootstrap-system-flowcontrol-configuration"
+
+type RESTStorageProvider struct{}
+
+var _ genericapiserver.PostStartHookProvider = RESTStorageProvider{}
+
+func (p RESTStorageProvider) NewRESTStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (genericapiserver.APIGroupInfo, bool, error) {
+	apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(flowcontrol.GroupName, legacyscheme.Scheme, legacyscheme.ParameterCodec, legacyscheme.Codecs)
+
+	if apiResourceConfigSource.VersionEnabled(flowcontrolv1alpha1.SchemeGroupVersion) {
+		apiGroupInfo.VersionedResourcesStorageMap[flowcontrolv1alpha1.SchemeGroupVersion.Version] = p.v1alpha1Storage(apiResourceConfigSource, restOptionsGetter)
+	}
+	return apiGroupInfo, true, nil
+}
+
+func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) map[string]rest.Storage {
+	storage := map[string]rest.Storage{}
+
+	// flow-schema
+	flowSchemaStorage, flowSchemaStatusStorage := flowschemastore.NewREST(restOptionsGetter)
+	storage["flowschemas"] = flowSchemaStorage
+	storage["flowschemas/status"] = flowSchemaStatusStorage
+
+	// priority-level-configuration
+	priorityLevelConfigurationStorage, priorityLevelConfigurationStatusStorage := prioritylevelconfigurationstore.NewREST(restOptionsGetter)
+	storage["prioritylevelconfigurations"] = priorityLevelConfigurationStorage
+	storage["prioritylevelconfigurations/status"] = priorityLevelConfigurationStatusStorage
+
+	return storage
+}
+
+func (p RESTStorageProvider) PostStartHook() (string, genericapiserver.PostStartHookFunc, error) {
+	predefined := PredefinedData{
+		FlowSchemas:                 flowcontrolbootstrap.PredefinedFlowSchemas(),
+		PriorityLevelConfigurations: flowcontrolbootstrap.PredefinedPriorityLevelConfigurations(),
+	}
+	return PostStartHookName, predefined.EnsurePredefinedConfiguration(), nil
+}
+
+type PredefinedData struct {
+	FlowSchemas                 []*flowcontrolv1alpha1.FlowSchema
+	PriorityLevelConfigurations []*flowcontrolv1alpha1.PriorityLevelConfiguration
+}
+
+func (d PredefinedData) EnsurePredefinedConfiguration() genericapiserver.PostStartHookFunc {
```

Comment:
(1) We are definitely thinking about what happens when an admin deletes (as well as creates or modifies) these things.  See the pending updates to the KEP for the intended behavior.  This PR has some bugs in these regards.

(2) As you say, this can be more often than we want, and we have no better place to do this.  I suspect the reason for the GET before the CREATE is to lighten then load when there are a lot of kube-apiservers.  I think an ideal solution would have a singleton controller whose jobs include initializing the cluster and setting status in the API objects to report problems with them.  I figured we would leave that for a later stage of development.

Reply:
currently in k/k, i see basically two way to deal with the preset configurations:

- in RBAC api, specifically `system:nodes` cluster-role, is written to the storage in the post-start-hook but never reload regardless of any future modifications on that object.
- in Scheduling api, all the preset objects are considered system-critical hence cannot be deleted/modified in the future.

in the beginning, i protected all these predefined FS/PL objects from deletion so that the system can be simpler and users are supposed to "mute" the undesired objects by modifying them to any invalid state. but @MikeSpreitzer suggests that somebody can hope to completely delete all the predefines to reset the world so now we allow the deletion... that's where the current implementation comes from i think..

Reply:
Alternatives like embedded controllers are difficult during HA upgrades, but I'm actually looking to fix one of my past sins where I should have used a controller instead of a post-starthook.

Reply:
The code here is specifically _not_ intended to maintain something forever; it is deliberately only for cluster initialization.  Plase look at the referenced KEP updates.